### ISSUE TITLE
Apply productions on specified subgraph

### ISSUE THREAD

**Issue by tomekzaw** [GramatykiGrafoweAGH/GramatykiGrafoweAGH#20]
*TL;DR we want to select where to apply the production instead of finding the subgraph automatically*

Currently, the productions are implemented as functions that accept an instance of graph and perform two steps:
1. programatically find the subgraph isomorphic to the left-hand side (LHS) of production that matches the applicability condition,
2. modify the graph inplace by adding, replacing, merging (or removing) nodes and edges.

If such subgraph cannot be found, then `CannotApplyProductionError` is raised. Productions mustn't modify the graph until we match the LHS of production and make sure that it can be applied.

For each production, we selected an arbitrary node in its LHS and call it the root node (i.e. `I` for P1-P6 or `E` for P7-P9). In order to find a subgraph isomorphic to LHS of production, we match the starting node first, and then we try to find the rest of the LHS of the production by matching the neighbours of already matched nodes.

Currently, the starting node is always matched with the "first" node with appropriate label:

https://github.com/GramatykiGrafoweAGH/GramatykiGrafoweAGH/blob/06d1b4827bb244fdaff6bd9559520c7d3de0d817/src/GramatykiGrafoweAGH/project/task1.py#L48

The above utility method always returns the node on the lowest level with the lowest auto-incremented ID:

https://github.com/GramatykiGrafoweAGH/GramatykiGrafoweAGH/blob/06d1b4827bb244fdaff6bd9559520c7d3de0d817/src/GramatykiGrafoweAGH/__init__.py#L133

Because of that, the productions are always applied on the lowest level possible, evenly expanding all branches of the graph (similarly to BFS). Otherwise, we would expand the graph unevenly (kinda similar to DFS) or even non-deterministically.

However, there are some use-cases where we would like to apply the production on the specified subgraph. Currently, it is not possible, as finding the subgraph isomorphic to LHS of production is implemented in the production function itself.

The plan is to divide these two steps into two separate functions/methods. The first function would only find the subgraph isomorphic to the LHS of the production (or raise `CannotApplyProductionError`). The latter, given the matched modes, would apply the production, i.e. modify the graph.

Here is the proposed interface:

```py
class IProduction(ABC):
    @abstractmethod
    def match(self, G: Graph) -> List[Node]:
        pass

    @abstractmethod
    def apply(self, G: Graph, nodes: List[Node]) -> None:
        pass
```

For example, P2 would be implemented as:

```py
class Production2(IProduction):
    def match(self, G: Graph) -> List[Node]:
        try:
            I = G.get_first_node_with_label('I')
            E1, E2, E3, E4 = get_square_vertices(G, I)
            return [I, E1, E2, E3, E4]
        except (NodeNotFoundError, SquareNotFoundError) as e:
            raise CannotApplyProductionError(e)

    def apply(self, G: Graph, nodes: List[Node]) -> None:
        I, E1, E2, E3, E4 = nodes
        # x1, y1 = E1.x, E1.y etc.
        # i = Node(label='i', x=I.x, y=I.y, level=level) etc.
        # modify the graph by adding nodes and edges
```

Here we use `List[Node]` to pass the matched nodes simply because of the fact that different productions will have different number of nodes in their LHS.

Please note that in this approach it is necessary to instantiate the production somewhere in the code, i.e. `G.apply_production(P2())`. This can lead to multiplying production instances unnecessarily. Another option would be to mark the methods as `@staticmethod`, but with non-static methods we could also parametrize the productions.

Here's a minimal implementation of `Graph.apply_production` method:

```py
def apply_production(self, production: Production) -> None:
    lhs = production.match(self)
    production.apply(self, lhs)
```

Finally, when visualizing the graph, we would like to apply a production each time we click a node. The clicked node becomes the root node. Then we match the rest of LHS of the production and finally modify the graph.

Here's the updated version of the proposed interface:

```py
class IProduction(ABC):
    @abstractmethod
    def find_root(self, G: Graph) -> Node:
        pass

    @abstractmethod
    def match_lhs(self, root: Node) -> List[Node]:
        pass

    @abstractmethod
    def apply(self, G: Graph, lhs: List[Node]) -> None:
        pass
```

Here's the example implementation for P2:

```py
class Production2(IProduction):
    def find_root(self, G: Graph) -> Node:
        try:
            return G.get_first_node_with_label('I')
        except NodeNotFoundError as e:
            raise CannotApplyProductionError(e)
    
    def match_lhs(self, root: Node) -> List[Node]:
        assert root.label == 'I'
        try:
            E1, E2, E3, E4 = get_square_vertices(G, root)
        except SquareNotFoundError as e:
            raise CannotApplyProductionError(e)
        return [root, E1, E2, E3, E4]

    def apply(self, G: Graph, lhs: List[Node]) -> None:
        I, E1, E2, E3, E4 = nodes
        # x1, y1 = E1.x, E1.y etc.
        # i = Node(label='i', x=I.x, y=I.y, level=level) etc.
        # modify the graph by adding nodes and edges
```

Assuming `root` is the clicked node, here's what we need to do to apply the production:

```py
lhs = production.match_lhs(root)
production.apply(G, lhs)
```

It is still possible to apply the production automatically, simply by calling:

```py
def apply_production(self, production: Production) -> None:
    root = production.find_root(G)
    lhs = production.match_lhs(root)
    production.apply(G, lhs)
```

But we still don't know which production should be applied. For this purpose, we would keep a list of productions ordered by the priority of applicability, starting from the most specific one (i.e. `productions = [P9, P8, P7, P6, P5, P4, P3, P2, P1]`). Each time a node is clicked, we iterate through the list and apply the first production that does not fail on `match_lhs(root)`. Here's the example implementation:

```py
def apply_production_on_clicked_node(root: Node) -> None:
    for production in productions:
        try:
            lhs = production.match_lhs(root)
        except CannotApplyProductionError:
            continue
        production.apply(G, lhs)
        return
    raise CannotApplyProductionError()
```

P.S. Yeah, I know, I'm a tryhard, but I really like this project (especially designing and implementing a developer-friendly API).
